File: 788_reverse_pair.py
def reverse_pair(ls):
    if len(ls) <= 1:
        return 0

    mid = len(ls) >> 1
    L = ls[:mid]
    R = ls[mid:]
    n = reverse_pair(L) + reverse_pair(R)

    i = j = k = 0
    while i < len(L) and j < len(R):
        if L[i] <= R[j]:
            ls[k] = L[i]
            i += 1
        else:
            ls[k] = R[j]
            j += 1
            n += len(L) - i  # 关键点
        k += 1
    while i < len(L):
        ls[k] = L[i]
        i += 1
        k += 1
    while j < len(R):
        ls[k] = R[j]
        j += 1
        k += 1
    return n


length = int(input())
ls = [int(x) for x in input().split()]
num = reverse_pair(ls)
print(num)

File: cpp_solution/788_reverse_pair.cpp
#include <iostream>

using namespace std;

const int N = 100010;

int q[N], tmp[N];

long reverse_pair(int q[], int l, int r) {
    if (l >= r) return 0;

    int mid = l + r >> 1;
    long num = reverse_pair(q, l, mid) + reverse_pair(q, mid + 1, r);
    
    int k = 0, i = l, j = mid + 1;
    while (i <= mid && j <= r) {
        if (q[i] <= q[j]) {
            tmp[k++] = q[i++];
        }
        else{
            tmp[k++] = q[j++];
            num += mid - i + 1;
        }
    }
    while (i <= mid) {
        tmp[k++] = q[i++];
    }
    while (j <= r) {
        tmp[k++] = q[j++];
    }
    for (int i = l, j = 0; i <= r; i++, j++) {
        q[i] = tmp[j];
    }
    return num;
}

int main() {
    int n;
    scanf("%d", &n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &q[i]);
    }
    printf("%ld", reverse_pair(q, 0, n - 1));
    return 0;
}